### PR TITLE
[Reviewer: AJH] Split storage configuration on Ralf

### PR DESCRIPTION
**Pull request by richardwhiuk**
Allow the Chronos cluster to be configured and pass the Ralf hostname
into the cluster as a callback URL.

This, along with the split-clusters work, allows Ralf to use an external storage node.

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -202,6 +208,11 @@ void usage(void)
        "                            The amount of time to blacklist a Diameter peer when it is unresponsive.\n"
        "     --astaire-blacklist-duration <secs>\n"
        "                            The amount of time to blacklist an Astaire node when it is unresponsive.\n"
+       "     --chronos-hostname <hostname>\n"
+       "                            The hostname of the remote Chronos cluster to use.\n"
+       "     --ralf-hostname <hostname>\n"
+       "                            The hostname of the cluster to which this Ralf is a member for the\n"
```
Comment (AlexHockey) [inline src/main.cpp]: I couldn't parse this sentence. It feels like one of the words is wrong.

File: src/main.cpp [hunk 2]
```diff
@@ -202,6 +208,11 @@ void usage(void)
        "                            The amount of time to blacklist a Diameter peer when it is unresponsive.\n"
        "     --astaire-blacklist-duration <secs>\n"
        "                            The amount of time to blacklist an Astaire node when it is unresponsive.\n"
+       "     --chronos-hostname <hostname>\n"
+       "                            The hostname of the remote Chronos cluster to use.\n"
```
Comment (AlexHockey) [inline src/main.cpp]: Say this defaults to localhost (logically at least).

File: src/main.cpp [hunk 3]
```diff
@@ -720,26 +739,49 @@ int main(int argc, char**argv)
...
+  std::string chronos_callback_addr;
+  int http_af;
+
+  if (options.chronos_hostname == "" || options.ralf_hostname == "")
   {
-    http_af = AF_INET6;
-    chronos_callback_addr = "[::1]:" + port_str;
-    local_chronos = "[::1]:7253";
+    // We want Chronos to call back to its local Ralf instance so that we can
+    // handle Ralfs failing without missing timers.
```
Comment (AlexHockey) [inline src/main.cpp]: Comment feels out of place in the new world.

File: debian/ralf.init.d [hunk 4]
```diff
@@ -135,6 +135,8 @@ get_daemon_args()
         [ -z "$cdf_identity" ] || billing_peer_arg="--billing-peer=$cdf_identity"
         [ -z $signaling_namespace ] || namespace_prefix="ip netns exec $signaling_namespace"
         [ -z "$local_site_name" ] || local_site_name_arg="--local-site-name=$local_site_name"
+        [ -z "$chronos_hostname" ] || chronos_hostname_arg="--chronos-hostname=$chronos_hostname"
+        [ -z "$ralf_hostname" ] || ralf_hostname_arg="--ralf-hostname=$chronos_hostname"
```
Comment (AlexHockey) [inline debian/ralf.init.d]: Unless I've misunderstood something, passing `chronos_hostname` twice is wrong.

File: src/main.cpp [hunk 5]
```diff
@@ -720,26 +739,49 @@ int main(int argc, char**argv)
...
+      chronos_callback_addr = "127.0.0.1:" + port_str;
+    }
+  }
+  else
+  {
+    chronos_service = options.chronos_hostname + ":7253";
+
+    http_af = is_ipv6(options.chronos_hostname) ? AF_INET6 : AF_INET;
+
+    chronos_callback_addr = options.ralf_hostname + ":" + port_str;
```
Comment (AlexHockey) [inline src/main.cpp]: `ralf_hostname` contains a port so doesn't this end up as `ralf.example.com:10888:10888`?

File: src/main.cpp [hunk 6]
```diff
@@ -720,26 +739,49 @@ int main(int argc, char**argv)
...
+    {
+      chronos_callback_addr = "127.0.0.1:" + port_str;
+    }
+  }
+  else
+  {
+    chronos_service = options.chronos_hostname + ":7253";
+
+    http_af = is_ipv6(options.chronos_hostname) ? AF_INET6 : AF_INET;
+
```
Comment (AlexHockey) [inline src/main.cpp]: Holy whitespace batman.

File: src/main.cpp (hunk 1 above)
Comment (AlexHockey) [inline src/main.cpp]: Worth mentioning that this will include the port.

**Comment by AlexHockey** [comment]
A few comments inline.

**Comment by richardwhiuk** [comment]
@eleanor-merry - I think you've separately implemented this on dev and merged that into split clusters, so this can be closed?

**Comment by eleanor-merry** [comment]
Yup

**Pull request closed without merging**